### ISSUE TITLE
When adding rows allow default data types to be overridden

### ISSUE THREAD

**Issue by cfsimplicity** [cfsimplicity/lucee-spreadsheet#200]
Currently when adding rows to workbooks, the POI cell data types are either auto-detected, or in the case of query data, derived from the query column types (unless specifically [ignored](https://github.com/cfsimplicity/lucee-spreadsheet/issues/199)).

Provide more granular control over the POI data types when adding rows by allowing the cell types in specified columns to be overridden.

Columns in data passed in as an array will need to be identified by number/position, but queries should also allow the column to be specified by name.

If the value to be set in a cell cannot be cast as the specified type, then the override should be ignored and fall back to the auto-detected or query column derived type instead. E.g. if a string/varchar column is forced to be a numeric column, but it contains values that can't be cast as numbers, those values should be typed as strings, not numbers.

See the [discussion in issue 199](https://github.com/cfsimplicity/lucee-spreadsheet/issues/199#issuecomment-676827286)

**Comment by cfsimplicity** [comment]
Implementation: add a new optional `datatypes` argument to the following methods:
```
addRow()
addRows()
binaryFromQuery()
downloadFileFromQuery()
workbookFromQuery()
writeFileFromQuery()
```
`datatypes` should be a struct containing up to 3 arrays, one for each of the possible override types: `string`, `numeric` and `date`.

Each array should contain the columns to which the type override should be applied. These can be either column numbers, or in the case of query data, column names. For example

```
// array data
data = [ [ "01234", 1234567890123456 ] ];
workbook = spreadsheet.newXlsx();
datatypes = { numeric: [ 1 ], string: [ 2 ] }; // you must use column numbers
spreadsheet.addRows( workbook=workbook, data=data, datatypes=datatypes );

// query data
data= QueryNew( "Col1,Col2,Col3", "VarChar,VarChar,BigInt", [ [ "01234", "2020-08-24", 1234567890123456 ] ] );
workbook = spreadsheet.newXlsx();
datatypes = { numeric: [ "Col1" ], date: [ "Col2" ], string: [ "Col3" ] }; // you can use column names
spreadsheet.addRows( workbook=workbook, data=data, datatypes=datatypes );
```

**Comment by JamoCA** [comment]
Are `boolean` and `time` supported?  (Just curious. I'm only asking because it wasn't included in the demo.)

It didn't appear to work as I thought it would... check this sample cfml out:
https://gist.github.com/JamoCA/05e8bd3b5adc758a878df1b67d88a1c7
- Default query column data types are all varchar.
- Without any override, the 1st column should remain strings (instead all are numeric)
- With the "date" override, the 2nd column should only have dates formatted as non-strings
- Row 9 (varchar) is converted to scientific notation (probably shouldn't)
- Row 11 varchar number should remain a string.
- Row 13 varchar ISO date string should remain a string.

**Comment by cfsimplicity** [comment]
Thanks for testing it out.

> Are boolean and time supported? (Just curious. I'm only asking because it wasn't included in the demo.)

No, ACF only supports `string`, `numeric` and `date` - I expect because they are the most likely to be needed - so for now I'm doing the same. I may look at adding them later.

> Without any override, the 1st column should remain strings (instead all are numeric)

Good catch. There was a hole in my logic which meant that query column types were being ignored for all columns even if they weren't being overridden. Fixed - please try again with the latest commit.

But your tests for rows 8-13 aren't valid as you haven't specified the `datatypes` option. Instead you're using the `ignoreQueryColumnDataTypes`, which only has an effect on *query* data, whereas you're adding data for these rows using arrays. Data added using arrays will always be auto typed unless you override them with `datatypes`.

> With the "date" override, the 2nd column should only have dates formatted as non-strings

Of the 7 varchar values you added as a query, only rows 4, 5 and 6 can be parsed as dates. If you use `getCellType( workbook, row, column )` you'll see that those 3 have been typed as `numeric` which is how Excel stores dates. The others fallback to varchar/string as expected.

> Row 9 (varchar) is converted to scientific notation (probably shouldn't)

Again, this is a POI/Excel limitation on numbers over 15 digits in length. Remove a digit and you'll see it's fine.

>Row 11 varchar number should remain a string.

It will if you pass it in a query varchar column and don't set `ignoreQueryColumnDataTypes`

> Row 13 varchar ISO date string should remain a string.

Why? It's a parseable date value, and if you've asked for that column to be overridden as dates then that's how it will be typed.

**Comment by JamoCA** [comment]
Array row autotypes without passing datatype override... that's important to remember.  Passing individual rows is not something I usually do, but I wanted to wrap my head around how the options worked together.  If the array data isn't part of the query, the newly added "row" has no awareness that it should use those types.  (My bad.)   I'll remember to pass datatypes via array if I ever use this method.

All query data works fine for me as of this update.

FYI:  I was creating some sub-tests and passed `9A` in an array to be autotyped.  Instead of a displaying `9A`, it was outputted as `09:00:00`.  I wasn't expecting this alphanumeric string to be converted to a time value. Upon further review, `TimeFormat("9A", "hh:mm:ss")` = `09:00:00`. So does `9{any number of spaces}A`.

**Comment by cfsimplicity** [comment]
No worries. It's easy to misunderstand when I haven't documented this feature properly yet.

It seems that time autotype quirk is ACF specific. Lucee won't accept `9A` as a date or time value. ACF will also treat `9p` as a time, although it outputs `09:00:00` when you'd expect `21:00:00`, so not consistent. I can make my internal "_IsDate()" function reject that pattern.

Let me know if you find any other problems, otherwise I'll aim for a new release in the next few days.

**Comment by JamoCA** [comment]
Thanks. I'll test this "9a" issue further and report it to Adobe as another bug.

What are your thoughts about adding an "auto" option to datatypes?  `datatypes={"auto"=["ColumnName"]}`  This way, the query column's datatype can override and fallback to whatever the cell is detected as by POI?

**Comment by cfsimplicity** [comment]
Do you mean the query column type would be ignored and the value auto detected? Obviously its use would be limited to query data where you haven't set `ignoreQueryColumnDataTypes`.

**Comment by JamoCA** [comment]
Yes, I believe it'd be beneficial to override VARCHAR at times to apply the autocasting logic to numeric/date data that clients have provided that is populated in VARCHAR columns.

**Comment by cfsimplicity** [comment]
OK, I've added support for "time", "boolean" and "auto" in the `datatypes` argument. Please test if you get chance.

**Comment by JamoCA** [comment]
Something's wrong now.  The datatype overrides appear to be ignored and all values are treated as the default query column type.  I cloned a query VARCHAR column and set up an override so it would be treated as "auto" and noticed that both columns that were to be overridden were ignored.

Sample CFML
https://gist.github.com/JamoCA/05e8bd3b5adc758a878df1b67d88a1c7

**Comment by cfsimplicity** [comment]
I tried your sample and it works as expected except for the 2 date/time values in rows 6 and 7, which were not accepted as date times and therefore left as strings:
```
8/22/2020 10:34 AM
12:53 pm
```
But I tried removing the "AM" and "pm" and they were then accepted as dates.

I'll take a look at the date time parsing with am/pm when I get chance. Everything else seems to be working fine though??

**Comment by cfsimplicity** [comment]
Found the problem: my regex in the fix for #201 wasn't specific enough and was blocking all date/times containing "am" or "pm" from being auto-detected or overridden.

Please try again with the latest `develop` commit.

**Comment by JamoCA** [comment]
This works correctly.  Thanks.
